### ISSUE TITLE
Passing body with manually defined 'Content-Length' produce invalid request

### ISSUE THREAD

**Issue by Ousret** [urllib3/urllib3#3045]
### Subject

Passing non-ascii 'str' as a body with manually defined 'Content-Length' produce an invalid request.
We are trusting the user input instead of internal mechanisms.
Should we allow trusting 'Content-Length' in that specific case? I know requests rely "incorrectly" on that behavior.
It was not detected until my experiment in #3030 because http.client is lax. h11 did raise an InternalProtocolError as expected.

Originally found while doing downstream **requests** tests.
`tests/test_requests.py::TestRequests::test_unicode_header_name`

### Environment

- urllib3 2.0.x
- urllib3 1.26.x

On any (supported) Python 3 interpreter or Operating system.

### Steps to Reproduce

```python
from urllib3 import request

if __name__ == "__main__":

    resp = request("POST", "http://localhost:3000", body="🚀", headers={"Content-Length": "1"})
```

### Expected Behavior

with fake server dumps.

```python
b"""POST / HTTP/1.1\r\nHost: localhost:3000\r\nAccept-Encoding: identity\r\nUser-Agent: python-urllib3/2.0.915\r\nContent-Length: 4\r\n\r\n\xf0\x9f\x9a\x80"""
```

### Actual Behavior

```python
b"""POST / HTTP/1.1\r\nHost: localhost:3000\r\nAccept-Encoding: identity\r\nUser-Agent: python-urllib3/2.0.915\r\nContent-Length: 1\r\n\r\n\xf0\x9f\x9a\x80"""
```

### Fix

Maybe an acceptable fix? https://github.com/urllib3/urllib3/commit/bd10e96f271e3762c9b161dc7bab133317da0e21#diff-542c95910a277028550d2c8943e8c49bbcb10f9af96e35d0a0eee99b8cfe9e8cR355

**Comment by sethmlarson** [comment]
Hmm, what happens when you pass bytes with an invalid Content-Length? I would be surprised if that had a different behavior.

**Comment by Ousret** [comment]
Good catch, you are absolutely right. The behavior is global.
So, let me rephrase, 

Passing `Content-Length` manually using a bad value tends to invalid data/requests sent across the socket.

**Comment by pquentin** [comment]
I'm not sure if I would do anything else than fixing the bug in requests itself here. Using h11 is going to flag a ton of things that are accepted today, but I'm not sure if we should inflict that pain onto users today.

**Comment by Ousret** [comment]
> but I'm not sure if we should inflict that pain onto users today.

Would you let people continue smoking if they say it would make them stress a lot to stop? :smile: 
I am not sure we should let users send garbage to innocent servers.

requests is rather in freeze mode, so I am not counting on patching anything.

**Comment by Ousret** [comment]
More generally, should we ban from input or..warn?:

- Content-Length
  - It is acceptable only when urllib3 cannot infer length by itself.
- Connection
- TE (trailers)
  - As I understand, not every HTTP expert like those, it's controversial. Some decide to ignore them. It may break the conn. and leave unconsummated bytes in the socket.
- Transfer-Encoding
- Keep-Alive

Those can break things and should be in sole urllib3 control.
This list may vary depending on the protocol version.

Ah, and yes, `requests` is aware that passing `str` may break things, it is documented on their end. https://requests.readthedocs.io/en/latest/user/advanced/#streaming-uploads

Maybe, if rephrased like this, I may understand best your interpretation of the state of the art.

Should **urllib3** guarantee to send a **valid** HTTP request through the pipes?
If not, we shall then close this issue :100: 

Thanks

**Comment by abebeos** [comment]
> Should **urllib3** guarantee to send a **valid** HTTP request through the pipes?

No, effort is too high.

> If not, we shall then close this issue

I'd say close.
